tree: Allow MapNodes from Records (#22042)
microsoft/FluidFramework: .changeset/green-spies-arrive.md

---
"fluid-framework": minor
"@fluidframework/tree": minor
---

Allow `Record` typed object to be used to construct MapNodes.

It is now allowed to construct MapNodes from `Record` typed objects, similar to how maps are expressed in JSON.

Before this change, an `Iterable<string, Child>` was required, but now an object like `{key1: Child1, key2: Child2}` is allowed.

Full example using this new API:
```typescript
class Schema extends schemaFactory.map("ExampleMap", schemaFactory.number) {}
const fromRecord = new Schema({ x: 5 });
```

> This new feature makes it possible for schema,

Reviewer: Rewording suggestion. I think I got the meaning right.
```suggestion
This new feature makes it possible for schemas to construct a tree entirely from JSON compatible objects using their constructors, as long as they do not require unhydrated nodes to differentiate ambiguous unions.
```

Author: If you want to suggest changing multiple lines, you should select multiple lines when making the comment. Applying that suggestion would just replace line 18 with that, which I'm pretty sure is not what you intended.

Author: applied slightly reworked version of change separate from this thread. Good rephrasing!

> which do not require unhydrated nodes to differentiate ambiguous unions,

Author: Detection of which schema have this ambiguity issue, and some related details and API which build on the Json support added in this PR are included in https://github.com/microsoft/FluidFramework/pull/22022 as drafts.

> This new feature makes it possible for schema,
> which do not require unhydrated nodes to differentiate ambiguous unions,
> to construct trees entirely from JSON compatible objects using their constructors.

Author: ```suggestion
This new feature makes it possible for schemas to construct a tree entirely from JSON compatible objects using their constructors,
as long as they do not require unhydrated nodes to differentiate ambiguous unions,
or IFluidHandles (which themselevs are not JSON compatible).
```

> Due to limitations of TypeScript and recursive types,
> recursive maps to not currently advertize support for this feature in their typing,
> but it does work at runtime.

Reviewer 2: ```suggestion
Due to limitations of TypeScript and recursive types,
recursive maps do not advertise support for this feature in their typing,
but it works at runtime.
```

Should maps be capitalized in this usage?
